DOCS-2553: Add Create a Module how-to (#3135)
viamrobotics/docs: docs/use-cases/create-module.md

---
title: "How to create and deploy a new module"
linkTitle: "Create a module"
type: "docs"
weight: 25
images: ["/registry/module-puzzle-piece.svg"]
tags: ["modular resources", "components", "services", "registry"]
description: "Add a custom resource by creating and deploying a module to your machine."
aliases:
  - /registry/create/
  - /registry/upload/
  - /extend/modular-resources/upload/
  - /modular-resources/upload/
---

{{<imgproc src="/registry/module-diagram.png" resize="x900" declaredimensions=true alt="Representation of the Viam registry, some modules within it, and a rover they support." >}}
<br>

> Viam provides built-in support for a variety of different {{< glossary_tooltip term_id="component" text="components" >}} and {{< glossary_tooltip term_id="service" text="services" >}}, but you can add support for your own [custom {{< glossary_tooltip term_id="resource" text="resources" >}}](/registry/) by creating a {{< glossary_tooltip term_id="module" text="module" >}}.
> 
> You can search the [Viam Registry](https://app.viam.com/registry) and [deploy an existing module to your machine](/registry/configure/) in a few clicks if you find one that meets your needs.

Reviewer: Shorten
```suggestion
Viam provides built-in support for a variety of different {{< glossary_tooltip term_id="component" text="components" >}} and {{< glossary_tooltip term_id="service" text="services" >}}, as well as a [Registry](https://app.viam.com/registry) full of modules.
```

Author: How about this?
```suggestion
Viam provides built-in support for a variety of different {{< glossary_tooltip term_id="component" text="components" >}} and {{< glossary_tooltip term_id="service" text="services" >}}, as well as a [registry](https://app.viam.com/registry) full of modules created by other users.
```

> If no existing modules support your specific use case, you can write your own module, and either upload it to the Viam registry to share with others, or deploy it to your machine as a local module without uploading it to the registry.

Reviewer: ```suggestion
If no existing modules support your specific use case, you can write your own [custom {{< glossary_tooltip term_id="resource" text="resources" >}}](/registry/) by creating a {{< glossary_tooltip term_id="module" text="module" >}}, and either upload it to the Viam registry to share it publicly, or deploy it to your machine as a local module without uploading it to the registry.
```

Author: ```suggestion
If no existing modules support your specific use case, you can write your own [custom {{< glossary_tooltip term_id="resource" text="resources" >}}](/registry/) by creating a module, and either upload it to the [Viam registry](https://app.viam.com/registry) to share it publicly, or deploy it to your machine as a local module without uploading it to the registry.
```

    If you do not intend to [upload your module](#upload-your-module-to-the-modular-resource-registry) to the [Viam registry](https://app.viam.com/registry), you do not need to use your organization's namespace as your model's namespace.
- `repo-name` is the code repository (GitHub repo) that houses your module code.
  - Ideally, your `repo-name` should describe the common functionality provided across the model or models of that module.
- `model-name` is the name of the new resource model that your module will provide.

For example, if your organization namespace is `acme`, and you have written a new base implementation named `mybase` which you have pushed to a repository named `my-custom-base-repo`, you would use the namespace `acme:my-custom-base-repo:mybase` for your model.

More requirements:

- Your model triplet must be all-lowercase.
- Your model triplet may only use alphanumeric (`a-z` and `0-9`), hyphen (`-`), and underscore (`_`) characters.

Determine the model name you want to use based on these requirements, then proceed to the next section.

> #### API namespace triplet and model namespace triplet examples

Author: Okay. Putting into an expander and moving it up a bit so people see it when they're confused by needing a second triplet :)

- The `rand:yahboom:arm` model and the `rand:yahboom:gripper` model use the repository name [yahboom](https://github.com/viam-labs/yahboom).
  The models implement the `rdk:component:arm` and the `rdk:component:gripper` API to support the Yahboom DOFBOT arm and gripper, respectively:

  ```json
  {
      "api": "rdk:component:arm",
      "model": "rand:yahboom:arm"
  },
  {
      "api": "rdk:component:gripper",
      "model": "rand:yahboom:gripper"
  }
  ```

- The `viam-labs:audioout:pygame` model uses the repository name [audioout](https://github.com/viam-labs/audioout).
  It implements the custom API `viam-labs:service:audioout`:

  ```json
  {
    "api": "viam-labs:service:audioout",
    "model": "viam-labs:audioout:pygame"
  }
  ```

The `viam` namespace is reserved for models provided by Viam.

> ### Write your new resource model definition

Author: Yes, it does feel overwhelming, but I just wanted to move this content over to here for now and then get into the details later. I edited the API triplet stuff quite a bit but not the model definition code stuff much at all.

At least one user indicates that examples are far and away the most useful thing, so whatever we do we want to make them as easy to find as possible. This could mean obvious prominent expanders, but yeah let's look at this in a separate PR?

};
```

{{< /expand >}}

{{% /tab %}}
{{< /tabs >}}

> #### (Optional) Configure logging

Reviewer: I think we should also do what Bill suggested and add a step where you write code that works first (without viam sdk) and then you wrap it into a module

> #### (Optional) Configure logging

Author: ^ I plan to do that step in some how-tos such as "How to write a sensor module with Python" but for this idk since it might not always be the way? Maybe instead, add it as a tip alert on this one?
